File: apps/alfred/schemas/today.py
from __future__ import annotations

from datetime import date
from typing import Any

from pydantic import BaseModel, Field


class TodayCaptureItem(BaseModel):
    id: str
    title: str
    source_url: str | None = None
    pipeline_status: str
    content_type: str | None = None
    created_at: str | None = None


class TodayStoredCardItem(BaseModel):
    card_id: int
    title: str
    topic: str | None = None
    status: str
    tags: list[str] = Field(default_factory=list)
    created_at: str | None = None


class TodayConnectionItem(BaseModel):
    link_id: int
    from_card_id: int
    from_title: str
    to_card_id: int
    to_title: str
    type: str
    created_at: str | None = None


class TodayReviewItem(BaseModel):
    review_id: int
    card_id: int
    card_title: str
    stage: int
    due_at: str | None = None
    completed_at: str | None = None
    status: str


class TodayGapItem(BaseModel):
    card_id: int
    title: str
    created_at: str | None = None


class TodayBriefingStats(BaseModel):
    total_captures: int = 0
    total_cards_created: int = 0
    total_connections: int = 0
    total_reviews_due: int = 0
    total_reviews_completed: int = 0
    total_gaps: int = 0
    total_events: int = 0
    total_cards: int = 0
    total_links: int = 0


class TodayBriefingResponse(BaseModel):
    date: str
    timezone: str
    generated_at: str
    captures: list[TodayCaptureItem] = Field(default_factory=list)
    stored_cards: list[TodayStoredCardItem] = Field(default_factory=list)
    connections: list[TodayConnectionItem] = Field(default_factory=list)
    reviews: list[TodayReviewItem] = Field(default_factory=list)
    gaps: list[TodayGapItem] = Field(default_factory=list)
    stats: TodayBriefingStats


class TodayCalendarDay(BaseModel):
    date: str
    captures: int = 0
    stored_cards: int = 0
    connections: int = 0
    reviews_due: int = 0
    reviews_completed: int = 0
    gaps: int = 0
    total_events: int = 0


class TodayCalendarResponse(BaseModel):
    start_date: str
    end_date: str
    timezone: str
    days: list[TodayCalendarDay] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# DailyEntry CRUD schemas (T3)
# ---------------------------------------------------------------------------


class DailyEntryCreate(BaseModel):
    """Create payload for a daily entry. ``kind`` is validated in the route
    layer (rejects ``artifact_ref``) so that Pydantic stays permissive."""

    entry_date: date
    kind: str
    title: str = Field(min_length=1, max_length=500)
    body_md: str = ""
    status: str = "open"
    priority: int = 0
    tags: list[str] = Field(default_factory=list)
    meta: dict = Field(default_factory=dict)
    user_id: str | None = None


class DailyEntryUpdate(BaseModel):
    """Partial patch; every field optional."""

    kind: str | None = None
    title: str | None = Field(default=None, min_length=1, max_length=500)
    body_md: str | None = None
    status: str | None = None
    priority: int | None = None
    tags: list[str] | None = None
    meta: dict | None = None
    entry_date: date | None = None


class DailyEntryItem(BaseModel):
    """Single entry as returned by ``GET /api/today/entries``.

    ``id`` is ``int`` for real rows and ``str`` (e.g. ``"zettel:123"``) for
    synthetic artifact-ref rows derived from zettels/captures/reviews.
    """

    id: int | str
    kind: str
    entry_date: str
    title: str
    body_md: str = ""
    status: str | None = None
    priority: int = 0
    tags: list[str] = Field(default_factory=list)
    meta: dict = Field(default_factory=dict)
    created_at: str | None = None
    updated_at: str | None = None
    is_synthetic: bool = False


class DailyEntriesResponse(BaseModel):
    entries: list[DailyEntryItem]
    next_cursor: str | None = None
    total: int = 0


# ---------------------------------------------------------------------------
# Daily reflection + manual pipeline trigger (T12)
# ---------------------------------------------------------------------------


class DailyReflectionResponse(BaseModel):
    """End-of-day digest returned by ``GET /api/today/reflections/{date}``."""

    id: int
    entry_date: str  # ISO date
    digest_md: str
    stats: dict
    pipeline_run_id: str
    stages_ran: list[str]
    generated_at: str
    user_id: str | None = None

    @classmethod
    def from_row(cls, row: Any) -> DailyReflectionResponse:
        """Build the response from a :class:`DailyReflectionRow`.

        Typed as ``Any`` to avoid a schema->model import cycle; the
        attribute access is structural and safe because the row shape is
        stable (see ``alfred.models.today.DailyReflectionRow``).
        """
        generated_at = getattr(row, "generated_at", None)
        return cls(
            id=row.id,
            entry_date=row.entry_date.isoformat(),
            digest_md=row.digest_md or "",
            stats=dict(row.stats or {}),
            pipeline_run_id=row.pipeline_run_id or "",
            stages_ran=list(row.stages_ran or []),
            generated_at=generated_at.isoformat() if generated_at else "",
            user_id=getattr(row, "user_id", None),
        )


class PipelineRunRequest(BaseModel):
    """Payload for ``POST /api/today/pipeline/run`` manual trigger."""

    entry_date: date
    tz: str = "UTC"
    user_id: str | None = None
    # ``False`` — synchronous execution for dev / small deployments.
    # ``True``  — dispatch to the Celery worker via ``.delay(...)``.
    enqueue: bool = False


class PipelineRunResponse(BaseModel):
    dispatched: bool
    task_id: str | None = None
    result: dict | None = None

File: web/lib/api/today.ts
import { ApiError, apiFetch, apiPatchJson, apiPostJson } from "@/lib/api/client";
import { apiRoutes } from "@/lib/api/routes";
import type {
  DailyEntriesResponse,
  DailyEntryCreate,
  DailyEntryItem,
  DailyEntryUpdate,
  DailyReflection,
  ListTodayEntriesParams,
  RunTodayPipelineBody,
  RunTodayPipelineResponse,
} from "@/features/today/types";

export type TodayCaptureItem = {
  id: string;
  title: string;
  source_url: string | null;
  pipeline_status: string;
  content_type: string | null;
  created_at: string | null;
};

export type TodayStoredCardItem = {
  card_id: number;
  title: string;
  topic: string | null;
  status: string;
  tags: string[];
  created_at: string | null;
};

export type TodayConnectionItem = {
  link_id: number;
  from_card_id: number;
  from_title: string;
  to_card_id: number;
  to_title: string;
  type: string;
  created_at: string | null;
};

export type TodayReviewItem = {
  review_id: number;
  card_id: number;
  card_title: string;
  stage: number;
  due_at: string | null;
  completed_at: string | null;
  status: string;
};

export type TodayGapItem = {
  card_id: number;
  title: string;
  created_at: string | null;
};

export type TodayNoteItem = {
  note_id: string;
  title: string;
  icon: string | null;
  workspace_id: string;
  updated_at: string | null;
};

export type TodayBriefingStats = {
  total_captures: number;
  total_cards_created: number;
  total_connections: number;
  total_reviews_due: number;
  total_reviews_completed: number;
  total_gaps: number;
  total_events: number;
  total_cards: number;
  total_links: number;
  total_notes_touched: number;
};

export type TodayBriefingResponse = {
  date: string;
  timezone: string;
  generated_at: string;
  captures: TodayCaptureItem[];
  stored_cards: TodayStoredCardItem[];
  connections: TodayConnectionItem[];
  reviews: TodayReviewItem[];
  gaps: TodayGapItem[];
  notes: TodayNoteItem[];
  stats: TodayBriefingStats;
};

export type TodayCalendarDay = {
  date: string;
  captures: number;
  stored_cards: number;
  connections: number;
  reviews_due: number;
  reviews_completed: number;
  gaps: number;
  total_events: number;
};

export type TodayCalendarResponse = {
  start_date: string;
  end_date: string;
  timezone: string;
  days: TodayCalendarDay[];
};

type TodayBriefingParams = {
  date?: string;
  tz?: string;
};

type TodayCalendarParams = {
  start_date?: string;
  end_date?: string;
  days?: number;
  tz?: string;
};

function buildQuery(params: Record<string, string | number | undefined>): string {
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined) continue;
    query.set(key, String(value));
  }
  return query.toString();
}

export async function getTodayBriefing(
  params: TodayBriefingParams = {},
): Promise<TodayBriefingResponse> {
  const query = buildQuery(params);
  const url = query ? `${apiRoutes.today.briefing}?${query}` : apiRoutes.today.briefing;
  return apiFetch<TodayBriefingResponse>(url, { cache: "no-store" });
}

export async function getTodayCalendar(
  params: TodayCalendarParams = {},
): Promise<TodayCalendarResponse> {
  const query = buildQuery(params);
  const url = query ? `${apiRoutes.today.calendar}?${query}` : apiRoutes.today.calendar;
  return apiFetch<TodayCalendarResponse>(url, { cache: "no-store" });
}

// ---------------------------------------------------------------------------
// DailyEntry CRUD — ``/api/today/entries`` (T3/T4)
// ---------------------------------------------------------------------------

function appendRepeated(
  query: URLSearchParams,
  key: string,
  values: readonly string[] | undefined,
): void {
  if (!values || values.length === 0) return;
  for (const value of values) {
    if (value === undefined || value === null) continue;
    query.append(key, String(value));
  }
}

function buildListEntriesQuery(params: ListTodayEntriesParams): string {
  const query = new URLSearchParams();
  query.set("start", params.start);
  query.set("end", params.end);
  if (params.tz) query.set("tz", params.tz);
  if (params.q) query.set("q", params.q);
  if (typeof params.include_artifacts === "boolean") {
    query.set("include_artifacts", params.include_artifacts ? "true" : "false");
  }
  if (typeof params.limit === "number") query.set("limit", String(params.limit));
  if (params.cursor) query.set("cursor", params.cursor);

  appendRepeated(query, "kind", params.kind);
  appendRepeated(query, "status", params.status);
  appendRepeated(query, "tag", params.tag);

  return query.toString();
}

export async function listTodayEntries(
  params: ListTodayEntriesParams,
): Promise<DailyEntriesResponse> {
  const query = buildListEntriesQuery(params);
  const url = query ? `${apiRoutes.today.entries}?${query}` : apiRoutes.today.entries;
  return apiFetch<DailyEntriesResponse>(url, { cache: "no-store" });
}

export async function createTodayEntry(body: DailyEntryCreate): Promise<DailyEntryItem> {
  return apiPostJson<DailyEntryItem, DailyEntryCreate>(apiRoutes.today.entries, body);
}

export async function updateTodayEntry(
  entryId: number,
  patch: DailyEntryUpdate,
): Promise<DailyEntryItem> {
  return apiPatchJson<DailyEntryItem, DailyEntryUpdate>(
    apiRoutes.today.entryById(entryId),
    patch,
  );
}

export async function deleteTodayEntry(entryId: number): Promise<void> {
  await apiFetch<unknown>(apiRoutes.today.entryById(entryId), { method: "DELETE" });
}

// ---------------------------------------------------------------------------
// Daily reflection + manual pipeline trigger (T12)
// ---------------------------------------------------------------------------

/**
 * Fetch the digest for a given date. Returns ``null`` for 404 (no
 * reflection yet) because the UI renders nothing in that case rather
 * than showing an error.
 */
export async function getTodayReflection(
  date: string,
  tz?: string,
): Promise<DailyReflection | null> {
  const base = apiRoutes.today.reflectionByDate(date);
  const url = tz ? `${base}?tz=${encodeURIComponent(tz)}` : base;
  try {
    return await apiFetch<DailyReflection>(url, { cache: "no-store" });
  } catch (error) {
    if (error instanceof ApiError && error.status === 404) {
      return null;
    }
    throw error;
  }
}

export async function runTodayPipeline(
  body: RunTodayPipelineBody,
): Promise<RunTodayPipelineResponse> {
  return apiPostJson<RunTodayPipelineResponse, RunTodayPipelineBody>(
    apiRoutes.today.pipelineRun,
    body,
  );
}
